### ISSUE TITLE
[RNN] Completely wrong predictions after dynamic range quantization

### ISSUE THREAD

**Issue by lucacampanella** [tensorflow/tensorflow#48774]
Hi everyone,  I'm using dynamic range quantization ([integer with float fallback](https://www.tensorflow.org/lite/performance/post_training_quantization#integer_with_float_fallback_using_default_float_inputoutput)) with a representative dataset, trying to resolved the prediction performance problem I highlighted in #48487.
Unfortunately the predictions I get from the model are wrong and don't correspond at all to the ones the non-quantized model outputs. The class probabilities are often all zeros, a part from one very close to 1. You can see an example of the output here:
```
tf.Tensor(
[[0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.99609375 0.         0.         0.
  0.         0.         0.        ]], shape=(1, 39), dtype=float32)
```

In order to make the conversion work, I have to set `unroll=True` when instantiating the model, while I have `unroll=False` during training. If I don't set `unroll=True`, then I get `Failed to parse the model: pybind11::init(): factory function returned nullptr.` (_Originally posted by @wenjielu123 in https://github.com/tensorflow/tensorflow/issues/39392#issuecomment-791634847_)

I'm handling the states myself, re-inputting the previous state in the next step.

I tried many different combinations of the possible flags, as explained in https://www.tensorflow.org/lite/performance/post_training_quantization . They all output strange predictions, with the full integer quantization outputting all zeros, but a couple of classes with the same values, for example 26.

Am I doing something wrong in the conversion? Why are the probabilities outputted by the model so off from the ones of the non-quantized model?

Thanks a lot for the help!
Luca

### 1. System information

OS Platform and Distribution: Linux Ubuntu 20.04
TensorFlow installation: pip package
TensorFlow library: tf-nightly==2.6.0.dev20210407

### 2. Code

Conversion code:
```
converter = tf.lite.TFLiteConverter.from_keras_model(model) 
converter.experimental_new_converter = True 
converter.experimental_new_quantizer = True 
converter.optimizations = [tf.lite.Optimize.DEFAULT] 
converter.representative_dataset = representative_dataset_generator 
# or also adding all the combinations of the following:
# converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8] 
# converter.inference_input_type = tf.uint8 
# converter.inference_output_type = tf.uint8
converter.convert() 
```

Model code:
```
def _build_test_graph(self, with_tf_lite_ready_setup=False):
        # Inputs
        if with_tf_lite_ready_setup:
            # Set the batch size of 1 when using TFLite, or it will result in a much bigger checkpoint that still only works with batch size 1
            batch_size = 1
            inputs = Input(shape=(1), batch_size=batch_size)
        else:
            batch_size = None
            inputs = Input(shape=(None,))

        # Input state
        h_0 = [Input(shape=[self.state_dim], batch_size=batch_size)]

        self.rnn_layer = tensorflow.keras.layers.GRU(self.state_dim, return_state=True, unroll=True)
        # unroll is set to false during training for simplicity, but to true during conversion, because otherwise
        # the conversion fails
        rnn_inputs = tf.one_hot(tf.cast(inputs, tf.int32), depth=self.vocab_size)

        embedder = Embedding(self.vocab_size, self.emb_dim, mask_zero=True)
        m = embedder.compute_mask(inputs) # the token 0 is the padding token
        # Apply RNN
        out = self.rnn_layer(rnn_inputs, initial_state=h_0, mask=m)

        rnn_out = out[0]
        h_out = out[1:]

        dense = Dense(self.vocab_size, activation="softmax")
        output = dense(rnn_out)

        return keras.Model([inputs] + h_0, [output] + h_out)
```

**Comment by abattery** [comment]
@jianlijianli could you triage this issue?

**Comment by jianlijianli** [comment]
Hi lucacampanella, thanks for creating the issue.

Some background: tensorflow lite offers several quantization schemes, including dynamic range quantization (only weights are quantized, activation is left float in the model and quantized on the fly during inference) and full integer quantization (both weights and activations are quantized; activation can only be quantized by running some calibration data through the "representative_dataset" API). Fully integer quantization is triggered by the "representative_dataset = ..." API. It's supposed to be less or equal accurate than dynamic range quantization but it gives better speed and can offload to DSPs.

In the current thread, you are using the full integer quantization by providing the representative_dataset. The flags you are using are reasonable and the full integer qunatization for GRU should be unfused. https://github.com/tensorflow/tensorflow/issues/48487 uses dynamic range quantization, which is supposed to have better (or equal) accuracy. If accuracy is the main concern, I think full integer quantization might not be the most sensible direction.

That being said, we need to figure out the issue in this thread. @liufengdb, can you please help take a look since this is on the new quantizer? Thanks.

**Comment by lucacampanella** [comment]
Hi @jianlijianli and thanks a lot for the answer, very useful. 

I had the intuition that using full integer quantization would yield worse accuracy than the dynamic range, as you are saying. I tried going down this road mostly to use the flag EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8, because in #48487 it was suggested that one difference with CoreML was the handling of activations. 

But if I understood correctly what you're saying, EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8 will also not help accuracy compared to the dynamic range, because in the best case the activations will be the same as the ones quantized on the fly by the dynamic range.

 So I guess this means there is no way to retain an accuracy similar to CoreML int8 with TFLite dynamic range because CoreML computes stuff in 16 bits and TFLite in 8 bits. Am I correct in this assumption?

Anyway, it could be nice to solve this conversion issue, let me know if you need any more details.

Thanks again!

**Comment by jianlijianli** [comment]
Hi lucacampanella,

Thanks for the discussion and sharing more details about your use case.

Didn't realized you are using EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8. With this extra flag, combined with "representative_dataset", the system performs a full integer quantization that has int8 weight and int16 activation (different from the int8 activation when EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8 is not set).

In theory, the accuracy of int16xint8 quantization sits in between int8xint8 ("int8 full integer quantization") and float32xint8 ("dynamic range quantization") and we will need to measure the accuracy to actually decide. (It could be possible int16xint8 is better than float32xint8)

One parameter for full integer quantization accuracy is the calibration dataset. If the representative_dataset is too small, it's not capturing all the essential min/max ranges so when a real data comes in, there will be clamping errors. On the other hand, if the representative_dataset is too large, it might captured some outlier ("outlier" is hard to quantify...) so the min/max range is unnecessarily large, which leads to wast of quantization space. I think we can try feed in more and less data to verify the accuracy.

Another possible approach is to use the [float16 quantization](https://www.tensorflow.org/lite/performance/post_training_float16_quant). Its accuracy should be close to float. It compress the model size by a factor of 2 but speed up is only observed in GPU.

**Comment by lucacampanella** [comment]
Hi jianlijianli,
Thanks a lot for the answer, makes it even clearer.

I tried different sizes for the representative dataset and none seems to be helping. The problem also presents itself in all the inputs, which maybe points to the fact that something is really off in the calibration.

Could the problem be that we consider the states as input? In order to keep the RNN "stateless" , we keep track of the states ourselves and fill the previous state as input to the next step to the model (as you can see from the code). In the representative dataset the state consists of floats, but when doing inference on the integer quantized model, the state will be of integers and may diverge amply from the states in the representative dataset?

Regarding float16, it works very well for us and maintains the same accuracy as the unquantized model. We're trying to reduce the size of the checkpoints of the model even further because we have one language model for every language supported by the app that needs to fit on the device :)

Thanks again!
